leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/LeshanServer.java

```diff
@@ -321,24 +282,57 @@ public LwM2mModelProvider getModelProvider() {
     @Override
     public <T extends LwM2mResponse> T send(Registration destination, DownlinkRequest<T> request)
             throws InterruptedException {
+        if (destination.usesQueueMode()) {
+            if (presenceService.isClientSleeping(destination)) {
+                LOG.info("The destination client is sleeping, request couldn't been sent.");
+                return null;
+            } else {
+                presenceService.startClientAwakeTimer(destination);
```

Comment:
A far as I have understood the LWM2M protocol, the client stays awake for `MAX_TRANXMIT_WAIT` after the last received request. So it will restart its internal timer on the reception of a request. So if we refresh the timer when we send the request, it will have a small drift with respect to the client's timer, due to the request transmit time. But if we restart it when we receive a response, it will have also a small drift due to the response transmit time, and the request processing time. So I think the question is what drift do we prefer? 

For me, it would be better to refresh it before sending the request, so that the server timer expires first and considers that the client is sleeping a little bit before it really is, rather than consider that it is sleeping a little bit after it really is.

Reply:
But this is more complicated than that because this is not just : 
```
Request ->  
<- Response  
```
In CoAP, there is request acknowledgement which could be separated of response and even worst you could imagine scenario with blockwise transfer that means 1 LWM2M request means several CoAP requests.
([How timeout is handled](https://github.com/eclipse/leshan/wiki/Request-Timeout) give an idea of the complexity)

In the spec, nothing say that the client should response in MAX_TRANSMIT_WAIT.
It must wait MAX_TRANSMIT_WAIT after the last request before to go to sleep but between a request and a response time could be longer.

Reply:
Ok I understand it, I've simplified it too much. Then we agree that the client will restart its timer when it gets a request.

But then when should the server restart this timer? I don't see any good solution because:
1. If restart when the coap request is ACK, wouldn't work for non-confirmable messages.
2. If restart when response, as you said the time between the request and the response could be much more larger than the actual awake time of the client.
So... any other idea?

Reply:
FMPOV, we will have difficulty to exactly know when the device is here or not..

So we should keep in mind this will just be **an approximation**.
For me this approximation must be good enough to allow users to sends requests to the client and not so more.

So, 2 solutions : 
1. We set the client awake on Response. Setting awake means change state + relaunch timer.
=> This means, in some cases we could consider the client sleeping although it is awake, but once we receive the response we will consider it awake again and so could send the next requests.
2. We pause the timer before to send the request and reactive it when we receive the response.
=> So we will not go in sleeping mode waiting we receive the response but this could be long.

Personally, I will go the 1. approximation as this is the simplest one and we will see later if this approximation cause troubles.

Reply:
I totally agree that this will be an approximation. I think the first option is also better because it feels more safe for me to consider that the client is sleeping when it is actually awake than the other way around.

So I guess this implementation needs to go in the next pull request where the Queue Mode sender is added so that we can add the timer restarting inside the onResponse of the message observer. So I think I'll just remove the restarting of the timer from here and we will discuss more about it in that PR, once this one is accepted. Do you agree with that?

Reply:
It sounds good to me :)